{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c64e25a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "import yaml\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pytorch_lightning import Trainer\n",
    "\n",
    "sys.path.append('..')\n",
    "sys.path.append('FrustraSeq')\n",
    "from FrustraSeq.models.FrustraSeq import FrustraSeq\n",
    "from FrustraSeq.dataloader import FrustrationDataModule"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f058066",
   "metadata": {},
   "source": [
    "# general info for script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b96679d8",
   "metadata": {},
   "source": [
    "## 1) Load Data\n",
    "First we load a simple fasta file containing the headers/ids and their sequences and store it in a dictionary of the form {id: sequence}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18bb1170",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file_path = \"../data/frustration/bonomi_ensembles_sequences.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "757a2cec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Alb3-A3CT': 'MDENASKIISAGRAKRSIAQPDDAGERFRQLKEQEKRSKKNKAVAKDTVELVEESQSESEEGSDDEEEEAREGALASSTTSKPLPEVGQRRSKRSKRKRTV',\n",
       " 'FCP1': 'PGPEEQEEEPQPRKPGTRRERTLGAPASSERSAAGGRGPRGHKRKLNEEDAASESSRESSNEDEGSSSEADEMAKALEAELNDLM',\n",
       " 'emerin_67-170': 'GTRGDADMYDLPKKEDALLYQSKGYNDDYYEESYFTTRTYGEPESAGPSRAVRQSVTSFPDADAFHHQVHDDDLLSSSEEECKDRERPMYGRDSAYQSITHYRPV',\n",
       " 'UBact': 'MIQSLMPERRERPGDPMPKSPSPLEEGGGPRRPETGSPDKDSLLKRMRRVDPKQAERYRQRTGE',\n",
       " 'Nsp2_CtlIDR': 'KEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGT',\n",
       " 'NHE1': 'MINNYLTVPAHKLDSPTMSRARIGSDPLAYEPKEDLPVITIDPASPQSPESVDLVNEELKGKVLGLSRDPAKVAEEDEDDDGGIMMRSKETSSPGTDDVFTPAPSDSPSSQRIQRCLSDP',\n",
       " 'p61_Hck': 'GGRSSCEDPGCPRDEERAPRMGCMKSKFLQVGGNTFSKTETSASPHCPVYVPDPTSTIKPGPNSHNSNTPGIREAGSE',\n",
       " 'ACTR': 'GTQNRPLLRNSLDDLVGPPSNLEGQSDERALLDQLHTLLSNTDATGLEEIDRALGIPELVNQGQALEPKQD',\n",
       " 'Hug1': 'AMADPMTMDQGLNPKQFFLDDVVLQDTLCSMSNRVNKSVKTGYLFPKDHVPSANIIAVERRGGLSDIGKNTSN',\n",
       " 'PaaA2': 'MDYKDDDDKNRALSPMVSEFETIEQENSYNEWLRAKVATSLADPRPAIPHDEVERRMAERFAKMRKERSKQ',\n",
       " 'Nt-SOCS5': 'RSLRQRLQDTVGLCFPMRTYSKQSKPLFSNKRKIHLSELMLEKCPFPAGSDLAQKWHLIKQHTAPVSPHS',\n",
       " 'Ab40': 'DAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV',\n",
       " 'Colicin_N_T_domain': 'GSNGADNAHNNAFGGGKNPGIGNTSGAGSNGSASSNRGNSNGWSWSNKPHKNDGFHSDGSYHITFHGDNNSKPKPGGNSGNRGNNGDGAS',\n",
       " 'asyn': 'MDVFMKGLSKAKEGVVAAAEKTKQGVAEAAGKTKEGVLYVGSKTKEGVVHGVATVAEKTKEQVTNVGGAVVTGVTAVAQKTVEGAGSIAAATGFVKKDQLGKNEEGAPQEGILEDMPVDPDNEAYEMPSEEGYQDYEPEA',\n",
       " 'idr_SSRP1': 'GSHMGYKDVDFGDSDNENEPDAYLARLKAEAREKEEDDDDGDSDEESTDEDFKPNENESDVAEEYDSNVESDSDDDSDASGGGGDSDGA',\n",
       " 'His-PknG_1-75': 'MGSSHHHHHHSSGLVPRGSHMAKASETERSGPGTQPADAQTATSATVRPLSTQAVFRPDFGDEDNFPHPTLGPDTEPQDRMATTSRVRPPVRRLG'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = {}\n",
    "with open(fasta_file_path, 'r') as f:\n",
    "    fasta_data = f.read()\n",
    "    for line in fasta_data.splitlines():\n",
    "        if line.startswith(\">\"):\n",
    "            header = line[1:]\n",
    "            seqs[header] = \"\"\n",
    "        else:\n",
    "            seqs[header] += line.strip()\n",
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1e695a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame.from_dict(seqs, orient='index', columns=[\"sequence\"]).reset_index().rename(columns={'index':'id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a5edb642",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alb3-A3CT</td>\n",
       "      <td>MDENASKIISAGRAKRSIAQPDDAGERFRQLKEQEKRSKKNKAVAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FCP1</td>\n",
       "      <td>PGPEEQEEEPQPRKPGTRRERTLGAPASSERSAAGGRGPRGHKRKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>emerin_67-170</td>\n",
       "      <td>GTRGDADMYDLPKKEDALLYQSKGYNDDYYEESYFTTRTYGEPESA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBact</td>\n",
       "      <td>MIQSLMPERRERPGDPMPKSPSPLEEGGGPRRPETGSPDKDSLLKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nsp2_CtlIDR</td>\n",
       "      <td>KEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                           sequence\n",
       "0      Alb3-A3CT  MDENASKIISAGRAKRSIAQPDDAGERFRQLKEQEKRSKKNKAVAK...\n",
       "1           FCP1  PGPEEQEEEPQPRKPGTRRERTLGAPASSERSAAGGRGPRGHKRKL...\n",
       "2  emerin_67-170  GTRGDADMYDLPKKEDALLYQSKGYNDDYYEESYFTTRTYGEPESA...\n",
       "3          UBact  MIQSLMPERRERPGDPMPKSPSPLEEGGGPRRPETGSPDKDSLLKR...\n",
       "4    Nsp2_CtlIDR      KEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGT"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c50ba01",
   "metadata": {},
   "source": [
    "## 2) Load Model\n",
    "We then load the checkpoint of our trained model and create a dataloader from the dictionary. Config and model weights can be found in the OneDrive folder (TODO update later with HF when published)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4941d72c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'it5_ABL_protT5_LORA'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load config.yaml corresponding to the model to be used.\n",
    "with open(f\"../data/it5_ABL_protT5_LORA/config.yaml\", 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "config[\"experiment_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41579682",
   "metadata": {},
   "outputs": [],
   "source": [
    "# either provide path to pretrained model weights or \n",
    "# set to huggingface model name (e.g. \"Rostlab/prot_t5_xl_uniref50\" for protT5) \n",
    "config[\"pLM_model\"] = \"Rostlab/prot_t5_xl_uniref50\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "91e12006",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using LoRA fine-tuning for ['q', 'k', 'v', 'o'] layers\n",
      "trainable params: 1,967,104 || all params: 1,210,107,904 || trainable%: 0.1626\n",
      "RANK -1: Model initialized.\n"
     ]
    }
   ],
   "source": [
    "model = FrustraSeq.load_from_checkpoint(checkpoint_path=f\"../data/{config['experiment_name']}/best_val_model.ckpt\",\n",
    "                                        config=config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3a4dd3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adjust dataloader parameters as needed. setting max_seq_length to the max length in the dataframe to avoid truncation.\n",
    "predict_dataloader = FrustrationDataModule(df=df,\n",
    "                                            max_seq_length=df[\"sequence\"].str.len().max(), \n",
    "                                            batch_size=5,\n",
    "                                            num_workers=1,\n",
    "                                            persistent_workers=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b138727b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mean': 0.24633155516241328, 'std': 0.5921655729624687}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adding the surprisal/discovery dictionary which is used to compute the surprisal feature during inference based on \n",
    "# precomputed values (for each aa) from the train set (found in onedrive as well)\n",
    "with open('../data/frustration/reg_heuristic.json', 'r') as f:\n",
    "    model.surprisal_dict = json.load(f)\n",
    "model.surprisal_dict[\"A\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "238602ef",
   "metadata": {},
   "source": [
    "## 3) Inference\n",
    "\n",
    "Lets define a Lightning trainer and run inference (prediction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e35b7e49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ’¡ Tip: For seamless cloud uploads and versioning, try installing [litmodels](https://pypi.org/project/litmodels/) to enable LitModelCheckpoint, which syncs automatically with the Lightning model registry.\n",
      "GPU available: True (mps), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n"
     ]
    }
   ],
   "source": [
    "trainer = Trainer(accelerator='mps',) # use 'gpu' instead of 'mps' on cuda enabled devices or 'cpu' for cpu only. mps is for Mac."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b4ddf1d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 16 sequences for prediction.\n",
      "Created test dataset for prediction\n",
      "Test dataset size: 16 samples\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:433: The 'predict_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=9` in the `DataLoader` to improve performance.\n",
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/torch/utils/data/dataloader.py:683: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, then device pinned memory won't be used.\n",
      "  warnings.warn(warn_msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting DataLoader 0:  25%|â–ˆâ–ˆâ–Œ       | 1/4 [00:04<00:14,  0.21it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/pytorch_lightning/loops/prediction_loop.py:257: predict returned None if it was on purpose, ignore this warning...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting DataLoader 0: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 4/4 [00:13<00:00,  0.29it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trainer.predict(model, predict_dataloader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "25af0f1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16,\n",
       " dict_keys(['residue', 'frustration_index', 'frustration_class', 'entropy', 'surprisal']))"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# model.pred_list contains a list of prediction dictionaries (one per protein)\n",
    "len(model.pred_list), model.pred_list[0].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf9edfe3",
   "metadata": {},
   "source": [
    "We then can either create protein specific dataframes or one combined one for all protein sequences in the input fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b6d51dc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>frustration_index</th>\n",
       "      <th>frustration_class</th>\n",
       "      <th>entropy</th>\n",
       "      <th>surprisal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>-0.300976</td>\n",
       "      <td>1</td>\n",
       "      <td>0.862157</td>\n",
       "      <td>-1.705295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D</td>\n",
       "      <td>0.093996</td>\n",
       "      <td>1</td>\n",
       "      <td>0.821093</td>\n",
       "      <td>0.657200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>0.624572</td>\n",
       "      <td>2</td>\n",
       "      <td>0.632576</td>\n",
       "      <td>1.546181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N</td>\n",
       "      <td>-0.455897</td>\n",
       "      <td>1</td>\n",
       "      <td>0.728476</td>\n",
       "      <td>0.109064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>0.224220</td>\n",
       "      <td>1</td>\n",
       "      <td>0.725651</td>\n",
       "      <td>-0.037340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  residue  frustration_index  frustration_class   entropy  surprisal\n",
       "0       M          -0.300976                  1  0.862157  -1.705295\n",
       "1       D           0.093996                  1  0.821093   0.657200\n",
       "2       E           0.624572                  2  0.632576   1.546181\n",
       "3       N          -0.455897                  1  0.728476   0.109064\n",
       "4       A           0.224220                  1  0.725651  -0.037340"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_protein_df = pd.DataFrame(model.pred_list[0]) # for the first protein\n",
    "per_protein_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ced04cfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>frustration_index</th>\n",
       "      <th>frustration_class</th>\n",
       "      <th>entropy</th>\n",
       "      <th>surprisal</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>-0.300976</td>\n",
       "      <td>1</td>\n",
       "      <td>0.862157</td>\n",
       "      <td>-1.705295</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D</td>\n",
       "      <td>0.093996</td>\n",
       "      <td>1</td>\n",
       "      <td>0.821093</td>\n",
       "      <td>0.657200</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>0.624572</td>\n",
       "      <td>2</td>\n",
       "      <td>0.632576</td>\n",
       "      <td>1.546181</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N</td>\n",
       "      <td>-0.455897</td>\n",
       "      <td>1</td>\n",
       "      <td>0.728476</td>\n",
       "      <td>0.109064</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>0.224220</td>\n",
       "      <td>1</td>\n",
       "      <td>0.725651</td>\n",
       "      <td>-0.037340</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1332</th>\n",
       "      <td>V</td>\n",
       "      <td>0.647613</td>\n",
       "      <td>2</td>\n",
       "      <td>0.617221</td>\n",
       "      <td>-1.649174</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1333</th>\n",
       "      <td>R</td>\n",
       "      <td>0.028138</td>\n",
       "      <td>1</td>\n",
       "      <td>0.743833</td>\n",
       "      <td>0.095156</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1334</th>\n",
       "      <td>R</td>\n",
       "      <td>0.043294</td>\n",
       "      <td>1</td>\n",
       "      <td>0.765684</td>\n",
       "      <td>0.113401</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335</th>\n",
       "      <td>L</td>\n",
       "      <td>0.656731</td>\n",
       "      <td>2</td>\n",
       "      <td>0.602273</td>\n",
       "      <td>-1.207965</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1336</th>\n",
       "      <td>G</td>\n",
       "      <td>-0.350341</td>\n",
       "      <td>1</td>\n",
       "      <td>0.384668</td>\n",
       "      <td>-0.041687</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1337 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     residue  frustration_index  frustration_class   entropy  surprisal  \\\n",
       "0          M          -0.300976                  1  0.862157  -1.705295   \n",
       "1          D           0.093996                  1  0.821093   0.657200   \n",
       "2          E           0.624572                  2  0.632576   1.546181   \n",
       "3          N          -0.455897                  1  0.728476   0.109064   \n",
       "4          A           0.224220                  1  0.725651  -0.037340   \n",
       "...      ...                ...                ...       ...        ...   \n",
       "1332       V           0.647613                  2  0.617221  -1.649174   \n",
       "1333       R           0.028138                  1  0.743833   0.095156   \n",
       "1334       R           0.043294                  1  0.765684   0.113401   \n",
       "1335       L           0.656731                  2  0.602273  -1.207965   \n",
       "1336       G          -0.350341                  1  0.384668  -0.041687   \n",
       "\n",
       "                 id  \n",
       "0         Alb3-A3CT  \n",
       "1         Alb3-A3CT  \n",
       "2         Alb3-A3CT  \n",
       "3         Alb3-A3CT  \n",
       "4         Alb3-A3CT  \n",
       "...             ...  \n",
       "1332  His-PknG_1-75  \n",
       "1333  His-PknG_1-75  \n",
       "1334  His-PknG_1-75  \n",
       "1335  His-PknG_1-75  \n",
       "1336  His-PknG_1-75  \n",
       "\n",
       "[1337 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_dfs = []\n",
    "for pred, id in zip(model.pred_list, df[\"id\"]):\n",
    "    pred[\"id\"] = id\n",
    "    pred_dfs.append(pd.DataFrame(pred))\n",
    "combined_df = pd.concat(pred_dfs, ignore_index=True)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbc96092",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving the results to csv files if wanted.\n",
    "per_protein_df.to_csv(\"./bonomi_protein1_predictions.csv\", index=False)\n",
    "combined_df.to_csv(\"./bonomi_all_proteins_predictions.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9511c5b8",
   "metadata": {},
   "source": [
    "Use entropy and surprisal score to filter trough predictions. Below is an example where want to filter for residues in which the model is confident (entropy >= 0.3) but its unlikely to observe this value given the amino acids frustration distribution (-1 >= surprisal score OR surprisal score > 1). A surprisal score of 1 means that the predicted regression value is one standart deviations away from its AA mean. Feel free to play around with both scores :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "65ec3e1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>frustration_index</th>\n",
       "      <th>frustration_class</th>\n",
       "      <th>entropy</th>\n",
       "      <th>surprisal</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>E</td>\n",
       "      <td>1.146633</td>\n",
       "      <td>2</td>\n",
       "      <td>0.343227</td>\n",
       "      <td>2.177570</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>R</td>\n",
       "      <td>1.218557</td>\n",
       "      <td>2</td>\n",
       "      <td>0.381284</td>\n",
       "      <td>1.528209</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>R</td>\n",
       "      <td>1.216357</td>\n",
       "      <td>2</td>\n",
       "      <td>0.369156</td>\n",
       "      <td>1.525560</td>\n",
       "      <td>emerin_67-170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>K</td>\n",
       "      <td>1.313926</td>\n",
       "      <td>2</td>\n",
       "      <td>0.307818</td>\n",
       "      <td>2.044763</td>\n",
       "      <td>emerin_67-170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>I</td>\n",
       "      <td>0.913598</td>\n",
       "      <td>2</td>\n",
       "      <td>0.177008</td>\n",
       "      <td>-1.189498</td>\n",
       "      <td>UBact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>E</td>\n",
       "      <td>1.247791</td>\n",
       "      <td>2</td>\n",
       "      <td>0.279468</td>\n",
       "      <td>2.299912</td>\n",
       "      <td>UBact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>D</td>\n",
       "      <td>-1.551834</td>\n",
       "      <td>0</td>\n",
       "      <td>0.353622</td>\n",
       "      <td>-1.125810</td>\n",
       "      <td>ACTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>D</td>\n",
       "      <td>1.263624</td>\n",
       "      <td>2</td>\n",
       "      <td>0.384908</td>\n",
       "      <td>1.924316</td>\n",
       "      <td>PaaA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>A</td>\n",
       "      <td>0.944697</td>\n",
       "      <td>2</td>\n",
       "      <td>0.261948</td>\n",
       "      <td>1.179341</td>\n",
       "      <td>PaaA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>808</th>\n",
       "      <td>E</td>\n",
       "      <td>1.356537</td>\n",
       "      <td>2</td>\n",
       "      <td>0.280731</td>\n",
       "      <td>2.431431</td>\n",
       "      <td>PaaA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>887</th>\n",
       "      <td>R</td>\n",
       "      <td>1.290334</td>\n",
       "      <td>2</td>\n",
       "      <td>0.379072</td>\n",
       "      <td>1.614615</td>\n",
       "      <td>Ab40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913</th>\n",
       "      <td>I</td>\n",
       "      <td>0.865705</td>\n",
       "      <td>2</td>\n",
       "      <td>0.182703</td>\n",
       "      <td>-1.324747</td>\n",
       "      <td>Ab40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>I</td>\n",
       "      <td>0.890624</td>\n",
       "      <td>2</td>\n",
       "      <td>0.092159</td>\n",
       "      <td>-1.254377</td>\n",
       "      <td>Ab40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>V</td>\n",
       "      <td>0.853249</td>\n",
       "      <td>2</td>\n",
       "      <td>0.048655</td>\n",
       "      <td>-1.038223</td>\n",
       "      <td>Ab40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>921</th>\n",
       "      <td>V</td>\n",
       "      <td>0.805221</td>\n",
       "      <td>2</td>\n",
       "      <td>0.106656</td>\n",
       "      <td>-1.180915</td>\n",
       "      <td>Ab40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>922</th>\n",
       "      <td>V</td>\n",
       "      <td>0.714947</td>\n",
       "      <td>2</td>\n",
       "      <td>0.139783</td>\n",
       "      <td>-1.449121</td>\n",
       "      <td>Ab40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1193</th>\n",
       "      <td>G</td>\n",
       "      <td>0.207372</td>\n",
       "      <td>1</td>\n",
       "      <td>0.379336</td>\n",
       "      <td>1.292433</td>\n",
       "      <td>idr_SSRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225</th>\n",
       "      <td>S</td>\n",
       "      <td>0.111866</td>\n",
       "      <td>1</td>\n",
       "      <td>0.392622</td>\n",
       "      <td>1.056185</td>\n",
       "      <td>idr_SSRP1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     residue  frustration_index  frustration_class   entropy  surprisal  \\\n",
       "32         E           1.146633                  2  0.343227   2.177570   \n",
       "36         R           1.218557                  2  0.381284   1.528209   \n",
       "188        R           1.216357                  2  0.369156   1.525560   \n",
       "268        K           1.313926                  2  0.307818   2.044763   \n",
       "292        I           0.913598                  2  0.177008  -1.189498   \n",
       "346        E           1.247791                  2  0.279468   2.299912   \n",
       "648        D          -1.551834                  0  0.353622  -1.125810   \n",
       "748        D           1.263624                  2  0.384908   1.924316   \n",
       "779        A           0.944697                  2  0.261948   1.179341   \n",
       "808        E           1.356537                  2  0.280731   2.431431   \n",
       "887        R           1.290334                  2  0.379072   1.614615   \n",
       "913        I           0.865705                  2  0.182703  -1.324747   \n",
       "914        I           0.890624                  2  0.092159  -1.254377   \n",
       "918        V           0.853249                  2  0.048655  -1.038223   \n",
       "921        V           0.805221                  2  0.106656  -1.180915   \n",
       "922        V           0.714947                  2  0.139783  -1.449121   \n",
       "1193       G           0.207372                  1  0.379336   1.292433   \n",
       "1225       S           0.111866                  1  0.392622   1.056185   \n",
       "\n",
       "                 id  \n",
       "32        Alb3-A3CT  \n",
       "36        Alb3-A3CT  \n",
       "188   emerin_67-170  \n",
       "268   emerin_67-170  \n",
       "292           UBact  \n",
       "346           UBact  \n",
       "648            ACTR  \n",
       "748           PaaA2  \n",
       "779           PaaA2  \n",
       "808           PaaA2  \n",
       "887            Ab40  \n",
       "913            Ab40  \n",
       "914            Ab40  \n",
       "918            Ab40  \n",
       "921            Ab40  \n",
       "922            Ab40  \n",
       "1193      idr_SSRP1  \n",
       "1225      idr_SSRP1  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.loc[(combined_df[\"entropy\"] <= 0.4) & ((combined_df[\"surprisal\"] <= -1) | (combined_df[\"surprisal\"] >= 1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7fa8fdc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mean': -0.050906470495787476, 'std': 0.8306870505452515}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change \"R\" to any amino acid single letter code to get its surprisal computation parameters\n",
    "model.surprisal_dict[\"R\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e539653",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biotrainer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
