{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c64e25a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import sys\n",
    "import yaml\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pytorch_lightning import Trainer\n",
    "\n",
    "sys.path.append('..')\n",
    "sys.path.append('pLMtrainer')\n",
    "from pLMtrainer.models.frustraSeq import FrustraSeq\n",
    "from pLMtrainer.dataloader import FrustrationDataModule"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f058066",
   "metadata": {},
   "source": [
    "# general info for script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b96679d8",
   "metadata": {},
   "source": [
    "## 1) Load Data\n",
    "First we load a simple fasta file containing the headers/ids and their sequences and store it in a dictionary of the form {id: sequence}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "18bb1170",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file_path = \"../data/frustration/bonomi_ensembles_sequences.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "757a2cec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Alb3-A3CT': 'MDENASKIISAGRAKRSIAQPDDAGERFRQLKEQEKRSKKNKAVAKDTVELVEESQSESEEGSDDEEEEAREGALASSTTSKPLPEVGQRRSKRSKRKRTV',\n",
       " 'FCP1': 'PGPEEQEEEPQPRKPGTRRERTLGAPASSERSAAGGRGPRGHKRKLNEEDAASESSRESSNEDEGSSSEADEMAKALEAELNDLM',\n",
       " 'emerin_67-170': 'GTRGDADMYDLPKKEDALLYQSKGYNDDYYEESYFTTRTYGEPESAGPSRAVRQSVTSFPDADAFHHQVHDDDLLSSSEEECKDRERPMYGRDSAYQSITHYRPV',\n",
       " 'UBact': 'MIQSLMPERRERPGDPMPKSPSPLEEGGGPRRPETGSPDKDSLLKRMRRVDPKQAERYRQRTGE',\n",
       " 'Nsp2_CtlIDR': 'KEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGT',\n",
       " 'NHE1': 'MINNYLTVPAHKLDSPTMSRARIGSDPLAYEPKEDLPVITIDPASPQSPESVDLVNEELKGKVLGLSRDPAKVAEEDEDDDGGIMMRSKETSSPGTDDVFTPAPSDSPSSQRIQRCLSDP',\n",
       " 'p61_Hck': 'GGRSSCEDPGCPRDEERAPRMGCMKSKFLQVGGNTFSKTETSASPHCPVYVPDPTSTIKPGPNSHNSNTPGIREAGSE',\n",
       " 'ACTR': 'GTQNRPLLRNSLDDLVGPPSNLEGQSDERALLDQLHTLLSNTDATGLEEIDRALGIPELVNQGQALEPKQD',\n",
       " 'Hug1': 'AMADPMTMDQGLNPKQFFLDDVVLQDTLCSMSNRVNKSVKTGYLFPKDHVPSANIIAVERRGGLSDIGKNTSN',\n",
       " 'PaaA2': 'MDYKDDDDKNRALSPMVSEFETIEQENSYNEWLRAKVATSLADPRPAIPHDEVERRMAERFAKMRKERSKQ',\n",
       " 'Nt-SOCS5': 'RSLRQRLQDTVGLCFPMRTYSKQSKPLFSNKRKIHLSELMLEKCPFPAGSDLAQKWHLIKQHTAPVSPHS',\n",
       " 'Ab40': 'DAEFRHDSGYEVHHQKLVFFAEDVGSNKGAIIGLMVGGVV',\n",
       " 'Colicin_N_T_domain': 'GSNGADNAHNNAFGGGKNPGIGNTSGAGSNGSASSNRGNSNGWSWSNKPHKNDGFHSDGSYHITFHGDNNSKPKPGGNSGNRGNNGDGAS',\n",
       " 'asyn': 'MDVFMKGLSKAKEGVVAAAEKTKQGVAEAAGKTKEGVLYVGSKTKEGVVHGVATVAEKTKEQVTNVGGAVVTGVTAVAQKTVEGAGSIAAATGFVKKDQLGKNEEGAPQEGILEDMPVDPDNEAYEMPSEEGYQDYEPEA',\n",
       " 'idr_SSRP1': 'GSHMGYKDVDFGDSDNENEPDAYLARLKAEAREKEEDDDDGDSDEESTDEDFKPNENESDVAEEYDSNVESDSDDDSDASGGGGDSDGA',\n",
       " 'His-PknG_1-75': 'MGSSHHHHHHSSGLVPRGSHMAKASETERSGPGTQPADAQTATSATVRPLSTQAVFRPDFGDEDNFPHPTLGPDTEPQDRMATTSRVRPPVRRLG'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs = {}\n",
    "with open(fasta_file_path, 'r') as f:\n",
    "    fasta_data = f.read()\n",
    "    for line in fasta_data.splitlines():\n",
    "        if line.startswith(\">\"):\n",
    "            header = line[1:]\n",
    "            seqs[header] = \"\"\n",
    "        else:\n",
    "            seqs[header] += line.strip()\n",
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e695a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame.from_dict(seqs, orient='index', columns=[\"sequence\"]).reset_index().rename(columns={'index':'id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a5edb642",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alb3-A3CT</td>\n",
       "      <td>MDENASKIISAGRAKRSIAQPDDAGERFRQLKEQEKRSKKNKAVAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FCP1</td>\n",
       "      <td>PGPEEQEEEPQPRKPGTRRERTLGAPASSERSAAGGRGPRGHKRKL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>emerin_67-170</td>\n",
       "      <td>GTRGDADMYDLPKKEDALLYQSKGYNDDYYEESYFTTRTYGEPESA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UBact</td>\n",
       "      <td>MIQSLMPERRERPGDPMPKSPSPLEEGGGPRRPETGSPDKDSLLKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nsp2_CtlIDR</td>\n",
       "      <td>KEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                           sequence\n",
       "0      Alb3-A3CT  MDENASKIISAGRAKRSIAQPDDAGERFRQLKEQEKRSKKNKAVAK...\n",
       "1           FCP1  PGPEEQEEEPQPRKPGTRRERTLGAPASSERSAAGGRGPRGHKRKL...\n",
       "2  emerin_67-170  GTRGDADMYDLPKKEDALLYQSKGYNDDYYEESYFTTRTYGEPESA...\n",
       "3          UBact  MIQSLMPERRERPGDPMPKSPSPLEEGGGPRRPETGSPDKDSLLKR...\n",
       "4    Nsp2_CtlIDR      KEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGT"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c50ba01",
   "metadata": {},
   "source": [
    "## 2) Load Model\n",
    "We then load the checkpoint of our trained model and create a dataloader from the dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4941d72c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'it5_ABL_protT5'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load config\n",
    "with open(f\"../data/it5_ABL_protT5/config.yaml\", 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "config[\"experiment_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41579682",
   "metadata": {},
   "outputs": [],
   "source": [
    "# either provide path to pretrained model or \n",
    "# set to huggingface model name (e.g. \"Rostlab/prot_t5_xl_uniref50\" for protT5-xl) \n",
    "config[\"pLM_model\"] = \"Rostlab/prot_t5_xl_uniref50\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "91e12006",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RANK -1: Model initialized.\n"
     ]
    }
   ],
   "source": [
    "model = FrustraSeq.load_from_checkpoint(checkpoint_path=f\"../data/{config['experiment_name']}/best_val_model.ckpt\",\n",
    "                                        config=config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3a4dd3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "predict_dataloader = FrustrationDataModule(df=df,\n",
    "                                            max_seq_length=df[\"sequence\"].str.len().max(),\n",
    "                                            batch_size=5,\n",
    "                                            num_workers=1,\n",
    "                                            persistent_workers=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b138727b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mean': 0.24633155516241328, 'std': 0.5921655729624687}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adding the surprisal dictionary which is used to compute the surprisal feature during inference based on \n",
    "# precomputed values (for each aa) from the train set\n",
    "with open('../data/frustration/reg_heuristic.json', 'r') as f:\n",
    "    model.surprisal_dict = json.load(f)\n",
    "model.surprisal_dict[\"A\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "238602ef",
   "metadata": {},
   "source": [
    "## 3) Inference\n",
    "\n",
    "Lets define a Lightning trainer and run inference (prediction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e35b7e49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ðŸ’¡ Tip: For seamless cloud uploads and versioning, try installing [litmodels](https://pypi.org/project/litmodels/) to enable LitModelCheckpoint, which syncs automatically with the Lightning model registry.\n",
      "GPU available: True (mps), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n"
     ]
    }
   ],
   "source": [
    "trainer = Trainer(accelerator='mps',) # use 'gpu' instead of 'mps' on cuda enabled devices or 'cpu' for cpu only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b4ddf1d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 16 sequences for prediction.\n",
      "Created test dataset for prediction\n",
      "Test dataset size: 16 samples\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:433: The 'predict_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=9` in the `DataLoader` to improve performance.\n",
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/torch/utils/data/dataloader.py:683: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, then device pinned memory won't be used.\n",
      "  warnings.warn(warn_msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting DataLoader 0:  25%|â–ˆâ–ˆâ–Œ       | 1/4 [00:03<00:10,  0.29it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/janleusch/anaconda3/envs/biotrainer/lib/python3.12/site-packages/pytorch_lightning/loops/prediction_loop.py:257: predict returned None if it was on purpose, ignore this warning...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting DataLoader 0: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 4/4 [00:10<00:00,  0.37it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None, None, None]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trainer.predict(model, predict_dataloader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "25af0f1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16,\n",
       " dict_keys(['residue', 'regression', 'classification', 'entropy', 'surprisal']))"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# model.pred_list contains a list of prediction dictionaries (one per protein)\n",
    "len(model.pred_list), model.pred_list[0].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf9edfe3",
   "metadata": {},
   "source": [
    "We then can either create protein specific dataframes or one combined one for all protein sequences in the input fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b6d51dc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>regression</th>\n",
       "      <th>classification</th>\n",
       "      <th>entropy</th>\n",
       "      <th>surprisal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>0.037727</td>\n",
       "      <td>1</td>\n",
       "      <td>0.978943</td>\n",
       "      <td>-1.093094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D</td>\n",
       "      <td>0.245120</td>\n",
       "      <td>1</td>\n",
       "      <td>0.908195</td>\n",
       "      <td>0.820920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>0.383294</td>\n",
       "      <td>2</td>\n",
       "      <td>0.811250</td>\n",
       "      <td>1.254376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N</td>\n",
       "      <td>0.279510</td>\n",
       "      <td>1</td>\n",
       "      <td>0.822857</td>\n",
       "      <td>1.229137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>0.415725</td>\n",
       "      <td>2</td>\n",
       "      <td>0.727638</td>\n",
       "      <td>0.286058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  residue  regression  classification   entropy  surprisal\n",
       "0       M    0.037727               1  0.978943  -1.093094\n",
       "1       D    0.245120               1  0.908195   0.820920\n",
       "2       E    0.383294               2  0.811250   1.254376\n",
       "3       N    0.279510               1  0.822857   1.229137\n",
       "4       A    0.415725               2  0.727638   0.286058"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_protein_df = pd.DataFrame(model.pred_list[0]) # for the first protein\n",
    "per_protein_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ced04cfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>regression</th>\n",
       "      <th>classification</th>\n",
       "      <th>entropy</th>\n",
       "      <th>surprisal</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M</td>\n",
       "      <td>0.037727</td>\n",
       "      <td>1</td>\n",
       "      <td>0.978943</td>\n",
       "      <td>-1.093094</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D</td>\n",
       "      <td>0.245120</td>\n",
       "      <td>1</td>\n",
       "      <td>0.908195</td>\n",
       "      <td>0.820920</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>0.383294</td>\n",
       "      <td>2</td>\n",
       "      <td>0.811250</td>\n",
       "      <td>1.254376</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N</td>\n",
       "      <td>0.279510</td>\n",
       "      <td>1</td>\n",
       "      <td>0.822857</td>\n",
       "      <td>1.229137</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>0.415725</td>\n",
       "      <td>2</td>\n",
       "      <td>0.727638</td>\n",
       "      <td>0.286058</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1332</th>\n",
       "      <td>V</td>\n",
       "      <td>0.972561</td>\n",
       "      <td>2</td>\n",
       "      <td>0.265225</td>\n",
       "      <td>-0.683747</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1333</th>\n",
       "      <td>R</td>\n",
       "      <td>-0.217082</td>\n",
       "      <td>1</td>\n",
       "      <td>0.572920</td>\n",
       "      <td>-0.200045</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1334</th>\n",
       "      <td>R</td>\n",
       "      <td>-0.014122</td>\n",
       "      <td>1</td>\n",
       "      <td>0.614342</td>\n",
       "      <td>0.044282</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335</th>\n",
       "      <td>L</td>\n",
       "      <td>0.976999</td>\n",
       "      <td>2</td>\n",
       "      <td>0.283272</td>\n",
       "      <td>-0.422705</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1336</th>\n",
       "      <td>G</td>\n",
       "      <td>-0.395529</td>\n",
       "      <td>1</td>\n",
       "      <td>0.452941</td>\n",
       "      <td>-0.149783</td>\n",
       "      <td>His-PknG_1-75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1337 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     residue  regression  classification   entropy  surprisal             id\n",
       "0          M    0.037727               1  0.978943  -1.093094      Alb3-A3CT\n",
       "1          D    0.245120               1  0.908195   0.820920      Alb3-A3CT\n",
       "2          E    0.383294               2  0.811250   1.254376      Alb3-A3CT\n",
       "3          N    0.279510               1  0.822857   1.229137      Alb3-A3CT\n",
       "4          A    0.415725               2  0.727638   0.286058      Alb3-A3CT\n",
       "...      ...         ...             ...       ...        ...            ...\n",
       "1332       V    0.972561               2  0.265225  -0.683747  His-PknG_1-75\n",
       "1333       R   -0.217082               1  0.572920  -0.200045  His-PknG_1-75\n",
       "1334       R   -0.014122               1  0.614342   0.044282  His-PknG_1-75\n",
       "1335       L    0.976999               2  0.283272  -0.422705  His-PknG_1-75\n",
       "1336       G   -0.395529               1  0.452941  -0.149783  His-PknG_1-75\n",
       "\n",
       "[1337 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_dfs = []\n",
    "for pred, id in zip(model.pred_list, df[\"id\"]):\n",
    "    pred[\"id\"] = id\n",
    "    pred_dfs.append(pd.DataFrame(pred))\n",
    "combined_df = pd.concat(pred_dfs, ignore_index=True)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dbc96092",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving the results to csv files if wanted.\n",
    "#per_protein_df.to_csv(\"./bonomi_protein1_predictions.csv\", index=False)\n",
    "combined_df.to_csv(\"./bonomi_all_proteins_predictions.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9511c5b8",
   "metadata": {},
   "source": [
    "Use entropy and surprisal score to filter trough predictions. Below is an example where want to filter for residues in which the model is confident (entropy >= 0.3) but its unlikely to observe this value given the amino acids frustration distribution (-1 >= surprisal score OR surprisal score > 1). A surprisal score of 1 means that the predicted regression value is one standart deviations away from its AA mean. Feel free to play around with both scores :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "65ec3e1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>residue</th>\n",
       "      <th>regression</th>\n",
       "      <th>classification</th>\n",
       "      <th>entropy</th>\n",
       "      <th>surprisal</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>R</td>\n",
       "      <td>0.956004</td>\n",
       "      <td>2</td>\n",
       "      <td>0.393493</td>\n",
       "      <td>1.212142</td>\n",
       "      <td>Alb3-A3CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>I</td>\n",
       "      <td>0.946722</td>\n",
       "      <td>2</td>\n",
       "      <td>0.230750</td>\n",
       "      <td>-1.095960</td>\n",
       "      <td>UBact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>Q</td>\n",
       "      <td>-0.172186</td>\n",
       "      <td>1</td>\n",
       "      <td>0.376972</td>\n",
       "      <td>1.037178</td>\n",
       "      <td>UBact</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>I</td>\n",
       "      <td>0.947645</td>\n",
       "      <td>2</td>\n",
       "      <td>0.321762</td>\n",
       "      <td>-1.093352</td>\n",
       "      <td>Nsp2_CtlIDR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>I</td>\n",
       "      <td>0.937512</td>\n",
       "      <td>2</td>\n",
       "      <td>0.378430</td>\n",
       "      <td>-1.121969</td>\n",
       "      <td>p61_Hck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964</th>\n",
       "      <td>G</td>\n",
       "      <td>0.148351</td>\n",
       "      <td>1</td>\n",
       "      <td>0.394822</td>\n",
       "      <td>1.151247</td>\n",
       "      <td>Colicin_N_T_domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>966</th>\n",
       "      <td>S</td>\n",
       "      <td>0.116753</td>\n",
       "      <td>1</td>\n",
       "      <td>0.362394</td>\n",
       "      <td>1.066665</td>\n",
       "      <td>Colicin_N_T_domain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    residue  regression  classification   entropy  surprisal  \\\n",
       "28        R    0.956004               2  0.393493   1.212142   \n",
       "292       I    0.946722               2  0.230750  -1.095960   \n",
       "350       Q   -0.172186               1  0.376972   1.037178   \n",
       "357       I    0.947645               2  0.321762  -1.093352   \n",
       "591       I    0.937512               2  0.378430  -1.121969   \n",
       "964       G    0.148351               1  0.394822   1.151247   \n",
       "966       S    0.116753               1  0.362394   1.066665   \n",
       "\n",
       "                     id  \n",
       "28            Alb3-A3CT  \n",
       "292               UBact  \n",
       "350               UBact  \n",
       "357         Nsp2_CtlIDR  \n",
       "591             p61_Hck  \n",
       "964  Colicin_N_T_domain  \n",
       "966  Colicin_N_T_domain  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.loc[(combined_df[\"entropy\"] <= 0.4) & ((combined_df[\"surprisal\"] <= -1) | (combined_df[\"surprisal\"] >= 1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa8fdc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mean': -0.050906470495787476, 'std': 0.8306870505452515}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change \"R\" to any amino acid single letter code to get its surprisal computation parameters\n",
    "model.surprisal_dict[\"R\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e539653",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biotrainer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
